### ISSUE TITLE
Stale IntelliSense squiggles 

### ISSUE THREAD

**Issue by bobbrow** [microsoft/vscode-cpptools#920]
(Copied from #915)
@Leedehai
The 2nd point - "Force Re-parsing" is ESPECIALLY a much needed feature! Sometimes the error squiggles still exist after fixing the problem, and these "unhinged" squiggles even move to a segment of code text that has no problem (the error a squiggle is used to underline no longer exists, so that squiggle becomes "unhinged").
I have to relaunch VSCode to git rid of them.

**Comment by bobbrow** [comment]
(Copied from #915)
@bobbrow
@Leedehai, if you are referring to the error squiggles from the new IntelliSense engine, we trigger a full reparse of the file after 10 seconds of inactivity, if you save the file, or if you change the editor focus to a different code file. The trigger resets the next time you edit the file. Do you still see squiggles after one of these triggers happens? I sincerely hope that you don't have to relaunch VS Code to clear them.

**Comment by bobbrow** [comment]
(Copied from #915)
@Leedehai
@bobbrow Hi thanks for your reply. Yes I am referring to the new IntelliSense engine.
The good news is, the triggers work (at least in my testing)! Thanks for pointing this out.

**Edit**: ..occasionally it still fails.. I tried to save the file or change focus to force a reparsing but the squiggle didn't go away. After relaunching VSCode it finally disappeared.
This squiggle indicated a "too few arguments" error, by the way, because I had a function call that does not match its declaration (which is in a .h file) and its definition (which is in the same .cpp file). My solution is not changing the function call, though, but change the function declaration AND definition.

**Comment by bobbrow** [comment]
I'd like to get more information about the case where it fails for you.  Can you outline the exact steps you take to reproduce the issue (including when you saved the .h or .cpp files during the process)?  I tried to emulate what you described, but I'm not getting stale squiggles.

As a last resort attempt to clear the squiggles, you still shouldn't need to relaunch VSCode.  You should only need to close the .cpp and .h files and reopen them.

**Comment by Leedehai** [comment]
@bobbrow Sorry it's hard to reproduce. It just happens occasionally. Just.. my feeling is that it is not as smooth as an IDE (yet). Good work, though:)

**Comment by xueyunquan** [comment]
@bobbrow : I don't know if this is relevant, but I also encounter "stale" squiggles from time to time. One thing that I've noticed is that after I added some include paths to the project setting file("c_cpp_properties.json") , under the "Linux" section, the corresponding #include statements are still shown with squiggles. The squiggles disappear only after I manually select the target configuration, by clicking the bottom-right part of the status bar and selecting "Linux" from the configuration drop list. Before I do that, the status bar shows "Linux" as the current configuration, but since the squiggles appear under the #include statements I know the true configuration in use must be another one. I'm using VS code 64-bit version in Ubuntu 17.04 x64.

**Comment by bobbrow** [comment]
@xueyunquan, I don't repro your case, though I know that there are potential issues with the file watchers on linux.  Did you save the changes to c_cpp_properties.json?  When the extension detects a change saved to this file, it discards all IntelliSense processes and starts over with the new settings.  If you have saved the changes to that file, then our extension must not be receiving the notification that the file changed.  I was thinking that we recently added some additional code to check the timestamp of that file periodically to work around that file watcher bug.  Do you have the latest version of the extension (0.12.1)?

**Comment by sean-mcmanus** [comment]
@xueyunquan If you go to Help->Toggle Developer Tools on Linux and you see a bunch of ENOSPC error, that means your OS ran out of file watchers. We added code to work around this for the c_cpp_properties.json, but it's possible there could be some additional issues (e.g. the call to fs.stat might be failing). There's information online for how to increase the number of file watchers on your OS if you want to try to work around the issue (assuming that's the cause).

**Comment by xueyunquan** [comment]
@bobbrow  Yeah, I have the latest 0.12.1 version. I always update VS code & extensions to the latest versions as soon as new ones available. 
And I always save the changes made to any setting files to make sure they take effects.

**Comment by xueyunquan** [comment]
@sean-mcmanus Yes, I do see lots of ENOSPC error. I'll try to increase the file watchers as you told. But I don't think it a nice idea to leave this to the users, no?

**Comment by sean-mcmanus** [comment]
@xueyunquan I don't think we should modify the user's OS settings, but if the workaround we have isn't working we should probably investigate more. From what I read online, other development tools that use file watchers have the same issue. See also https://code.visualstudio.com/docs/setup/linux#_error-enospc .

**Comment by xueyunquan** [comment]
@sean-mcmanus @bobbrow : Thanks for your reply & info. I'd like to put more feedbacks and information in the future.

**Comment by bobbrow** [comment]
I'm closing old issues marked "more info needed".  If you are still having issues with the extension, please ensure you have installed the latest version and open a new issue with the issue details.
